### PR TITLE
Define boundary conditionts 9

### PR DESCRIPTION
**Pull request by morettid**

### CODE REVIEW

**Comment by morettid** [comment]
Conditional to eventual comments on pull request #32, this pull request is ready for review.

If any fixes are needed for pull request #32, those can be propagated to this pull request #34.

File: stem/boundary.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        return True
+
+@dataclass
+class AbsorbingBoundary(BoundaryParametersABC):
+    """
+    Class containing the boundary parameters for a point boundary
+
+    Attributes:
+        absorbing_factors (List[float]): Activate/deactivate boundary for each direction.
+        virtual_thickness (float): Entity of the boundary in the 3 directions [N].
```
Comment (aronnoordam) [inline stem/boundary.py:68]: the virtual thickness is 1-dimensional

File: stem/boundary.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    def is_constraint(self) -> bool:
+        return True
+
+@dataclass
+class AbsorbingBoundary(BoundaryParametersABC):
+    """
+    Class containing the boundary parameters for a point boundary
+
+    Attributes:
+        absorbing_factors (List[float]): Activate/deactivate boundary for each direction.
```
Comment (aronnoordam) [inline stem/boundary.py:67]: abosrbing factors are not no activate the boundary, These indicate how much of the P-wave and S-wave should be damped

Comment (aronnoordam) [reply stem/boundary.py:67]: 0, is no damping; 1: full damping; 0<x,1: partial damping

File: stem/IO/kratos_boundaries_io.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            boundary (Load): boundary object
+
+        Returns:
+            Dict[str, Any]: dictionary containing the boundary parameters
+        """
+
+        # add boundary parameters to dictionary based on boundary type.
+
+        if isinstance(boundary.boundary_parameters, DisplacementConstraint):
+            return self.__create_displacement_constraint_dict(boundary=boundary)
```
Comment (aronnoordam) [inline stem/IO/kratos_boundaries_io.py:120]:
```suggestion
            return self.__create_displacement_constraint_dict(boundary)
```

same for the others

File: stem/IO/kratos_boundaries_io.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        boundary_dict["Parameters"][
+            "model_part_name"
+        ] = f"{self.domain}.{boundary.part_name}"
+        boundary_dict["Parameters"]["variable_name"] = "ROTATION"
+        boundary_dict["Parameters"]["table"] = [0, 0, 0]
+
+        return boundary_dict
+
+    def __create_absorbing_boundary_dict(self, boundary: Boundary) -> Dict[str, Any]:
```
Comment (aronnoordam) [inline stem/IO/kratos_boundaries_io.py:81]: the typing can be AbsorbingBoundary, same for the others

File: stem/IO/kratos_boundaries_io.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        constraints_dict: Dict[str, Any] = {"constraints_process_list": []}
+        absorbing_boundaries_list: List[Dict[str, Any]] = []
+
+        for boundary in boundaries:
+            boundary_dict = self.__create_boundary_dict(boundary)
+            if boundary.boundary_parameters.is_constraint():
+                constraints_dict["constraints_process_list"].append(boundary_dict)
+            else:
+                absorbing_boundaries_list.append(boundary_dict)
```
Comment (aronnoordam) [inline stem/IO/kratos_boundaries_io.py:154]: even though absorbing boundaries are currently the only non-constraints boundaries, it might change in the future, therefore i would rename this to load_boundaries

File: stem/IO/kratos_boundaries_io.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    def create_dictionaries_for_boundaries(
+        self, boundaries: List[Boundary]
+    ) -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
+        """
+        Creates a dictionary containing the `constraint_process_list` (list of
+        dictionaries to specify the constraints for the model) and a list of
+        dictionaries for the absorbing boundaries to be given to `load_process_list`
+
+        Args:
+            boundaries (List[Boundary]): list of load objects
```
Comment (aronnoordam) [inline stem/IO/kratos_boundaries_io.py:137]: naming is incorrect

File: stem/boundary.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+@dataclass
+class RotationConstraint(BoundaryParametersABC):
+    """
+    Class containing the boundary parameters for rotation constraint
+
+    Attributes:
+        active (List[bool]): Activate/deactivate constraint for each direction.
+        is_fixed (List[bool]): Specify if constraint is fixed.
+        value (List[float]):
```
Comment (aronnoordam) [inline stem/boundary.py:51]: rotation constraint around x,y and z axis

File: stem/boundary.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+
+@dataclass
+class RotationConstraint(BoundaryParametersABC):
+    """
+    Class containing the boundary parameters for rotation constraint
+
+    Attributes:
+        active (List[bool]): Activate/deactivate constraint for each direction.
+        is_fixed (List[bool]): Specify if constraint is fixed.
```
Comment (aronnoordam) [inline stem/boundary.py:50]: around each axis

File: stem/boundary.py (hunk 2 above)
Comment (morettid) [reply stem/boundary.py:67]: Thanks for the explanation, will include it in the docstring!

File: stem/IO/kratos_boundaries_io.py (hunk 3 above)
Comment (morettid) [reply stem/IO/kratos_boundaries_io.py:120]: Currently changed due to change in signature of the inputs (part_name, boundary_parameters).
`boundary` is now used as self.__create_displacement_constraint_dict(**boundary.__dict__).

File: stem/boundary.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    """
+
+    active: List[bool]
+    is_fixed: List[bool]
+    value: List[float]
+
+    @property
+    def is_constraint(self) -> bool:
+        """
+        Determines whether the boundary should be in the list of
```
Comment (aronnoordam) [inline stem/boundary.py:47]: since is_constraint is a property and is always true for the displacementconstraint. Can you indicate that in the docstring. E.g.: " Property which indicates if boundary condition is a constraint. True for  DisplacementConstraint "

same for the rest of the is_constraint defs

File: stem/boundary.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    Abstract base class for boundary parameters
+    """
+
+    pass
+
+    @property
+    @abstractmethod
+    def is_constraint(self) -> bool:
+        """
```
Comment (aronnoordam) [inline stem/boundary.py:17]: Property which indicates if boundary condition is a constraint. 
 
 The fact that we later use it to check if it should be in a list of loads or constraints, is irrelevant for the user. Thats a pure Kratos IO thing

**Comment by morettid** [comment]
@aronnoordam I am implementing the changes in a new pull request in which I am a bit refracturing the repo.

Will make an intermediate pull request soon enough, but this pull request is superseded.

**Comment by aronnoordam** [comment]
> @aronnoordam I am implementing the changes in a new pull request in which I am a bit refracturing the repo.
> 
> Will make an intermediate pull request soon enough, but this pull request is superseded.

I don't completely understand. Do you mean that I have to wait reviewing this PR until the other PR is merged?

File: stem/boundary.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        Property which indicates if boundary condition is a constraint.
+        Returns:
+            bool
+        """
+        pass
```
Comment (aronnoordam) [inline stem/boundary.py:22]:
```suggestion
        raise Exception("abstract method 'is_constraint' of boundary parameters class is called")
```

File: stem/boundary.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        Property which indicates if boundary condition is a constraint. False for AbsorbingBoundary.
+        Returns:
+            bool
+        """
+        return False
+
+
+@dataclass
+class Excavation(BoundaryParametersABC):
```
Comment (aronnoordam) [inline stem/boundary.py:110]: excavation is not a boundary condition. I think this belongs in a separate file, something like: additional_processes.py

File: stem/default_materials.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    TwoPhaseSoil,
+    LinearElasticSoil,
+    SaturatedBelowPhreaticLevelLaw,
+    FluidProperties,
+)
+from stem.structural_material import StructuralMaterial, EulerBeam
+
+
+def default_peat_material() -> SoilMaterial:
+    return SoilMaterial(
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/default_materials.py:13]: also all the default soil properties have the same parameters. This is not realistic, I think this has to be thought about, but maybe belongs in a different pull request

File: stem/load.py [hunk 14]
```diff
@@ -53,23 +79,18 @@ class MovingLoad(LoadParametersABC):
     offset: float = 0.0
 
 
-class Load:
+@dataclass
+class GravityLoad(LoadParametersABC):
     """
-    Class containing load information acting on a body part
+    Class containing the load parameters for a gravity loads.
```
Comment (aronnoordam) [inline stem/load.py:85]:
```suggestion
    Class containing the load parameters for a gravity load.
```

File: stem/boundary.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    """
+    Class containing the boundary parameters for displacement constraint
+
+    Inheritance:
+        - :class:`BoundaryParametersABC`
+
+    Attributes:
+        active (List[bool]): Activate/deactivate constraint for each direction.
+        is_fixed (List[bool]): Specify if constraint is fixed for each direction.
+        value (List[float]): Displacement constraint for direction [m].
```
Comment (aronnoordam) [inline stem/boundary.py:36]:
```suggestion
        - active (List[bool]): Activate/deactivate constraint for each direction.
        - is_fixed (List[bool]): Specify if constraint is fixed for each direction.
        - value (List[float]): Displacement constraint for direction [m].
```

please correct all docstrings in the PR with the new docstring style, such that readthedocs will be generated nicely

File: stem/default_materials.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        retention_parameters=SaturatedBelowPhreaticLevelLaw(),
+        fluid_properties=FluidProperties(
+            DENSITY_FLUID=1.000,
+            DYNAMIC_VISCOSITY=8.9e-7,
+            BULK_MODULUS_FLUID=2.0e-30,
+        ),
+    )
+
+
+def default_steel_beam_material(ndim:int) -> StructuralMaterial:
```
Comment (aronnoordam) [inline stem/default_materials.py:81]: missing docstring

File: stem/default_materials.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+                YOUNG_MODULUS=2.1e11,
+                POISSON_RATIO=0.2,
+                CROSS_AREA=1.0,
+                I33=1
+            ),
+        )
+    else:
+        raise ValueError
+
+class DefaultMaterial(Enum):
```
Comment (aronnoordam) [inline stem/default_materials.py:111]: missing docstring

File: stem/default_materials.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+            DYNAMIC_VISCOSITY=8.9e-7,
+            BULK_MODULUS_FLUID=2.0e-30,
+        ),
+    )
+
+
+def default_steel_beam_material(ndim:int) -> StructuralMaterial:
+    if ndim == 3:
+        return StructuralMaterial(
+            name="default_elastic_steel_beam_3D",
```
[the remark this answers is not in the document]
Comment (morettid) [reply stem/default_materials.py:84]: Found them in here: https://www.valente1919.com/rails/ and assumed J_zz for the torsional stiffness (I know it's not exact, but is my best guess atm).

File: stem/default_materials.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    Currently, the most common rails adopted in the Netherlands are implemented (46E3, 54E1 and 60E1).
+
+    Args:
+        - ndim (int): number of dimensions of the problem (either 2 or 3)
+        - rail_type (:class:`RailTypes`): instance of the enumeration to describe the rail type
+
+    Returns:
+        - :class:`stem.structural_material.StructuralMaterial`
+    """
+    if rail_type.value == 1:
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/default_materials.py:30]: I think checking the whole type makes it a bit clearer what's going on

File: stem/IO/kratos_loads_io.py [hunk 20]
```diff
@@ -62,48 +64,113 @@ def __create_moving_load_dict(self, load: Load) -> Dict[str, Any]:
...
+    def __create_surface_load_dict(self, part_name:str, parameters: LoadParametersABC) -> Dict[str, Any]:
+        """
+        Creates a dictionary containing the surface load parameters
+
+        Args:
+            - part_name (str): name of the model part on which the load is applied
+            - parameters (:class:`stem.load.SurfaceLoad`): surface load parameters object
+
+        Returns:
+            Dict[str, Any]: dictionary containing the load parameters
```
Comment (aronnoordam) [inline stem/IO/kratos_loads_io.py:110]:
```suggestion
            - Dict[str, Any]: dictionary containing the load parameters
```

File: stem/boundary.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+
+    pass
+
+    @property
+    @abstractmethod
+    def is_constraint(self) -> bool:
+        """
+        Property which indicates if boundary condition is a constraint.
+        Returns:
+            bool
```
Comment (aronnoordam) [inline stem/boundary.py:20]:
```suggestion
        Property which indicates if boundary condition is a constraint.
        
        Returns:
            - bool
```

**Review by aronnoordam — COMMENTED** [review]
few suggestions, and a test for default materials is missing

**Review by aronnoordam — APPROVED** [review]
